cupy/fft/_callback.pyx

```diff
@@ -467,53 +498,215 @@ cdef class _CallbackManager:
+        ``cufftXtJITSetCallback``.
 
         .. note::
-            After this method is called, a call to :meth:`set_callbacks` must
-            follow. This is for internal use.
+            If :meth:`set_caller_infos` is called, a call to this method must
+            follow.
 
         '''
-        self.cb_load_aux_arr = cb_load_aux_arr
-        self.cb_store_aux_arr = cb_store_aux_arr
+        from cupy.cuda.cufft import (
+            CUFFT_C2C, CUFFT_C2R, CUFFT_R2C,
+            CUFFT_Z2Z, CUFFT_Z2D, CUFFT_D2Z,
+            CUFFT_CB_LD_COMPLEX, CUFFT_CB_LD_COMPLEX_DOUBLE,
+            CUFFT_CB_LD_REAL, CUFFT_CB_LD_REAL_DOUBLE,
+            CUFFT_CB_ST_COMPLEX, CUFFT_CB_ST_COMPLEX_DOUBLE,
+            CUFFT_CB_ST_REAL, CUFFT_CB_ST_REAL_DOUBLE,)
+
+        cdef MemoryPointer cb_load_data = self.cb_load_data
+        cdef MemoryPointer cb_store_data = self.cb_store_data
+        cdef intptr_t cb_load_ptr=0, cb_store_ptr=0
+
+        if fft_type == CUFFT_C2C:
+            cb_load_type = CUFFT_CB_LD_COMPLEX if self.cb_load else -1
+            cb_store_type = CUFFT_CB_ST_COMPLEX if self.cb_store else -1
+        elif fft_type == CUFFT_R2C:
+            cb_load_type = CUFFT_CB_LD_REAL if self.cb_load else -1
+            cb_store_type = CUFFT_CB_ST_COMPLEX if self.cb_store else -1
+        elif fft_type == CUFFT_C2R:
+            cb_load_type = CUFFT_CB_LD_COMPLEX if self.cb_load else -1
+            cb_store_type = CUFFT_CB_ST_REAL if self.cb_store else -1
+        elif fft_type == CUFFT_Z2Z:
+            cb_load_type = CUFFT_CB_LD_COMPLEX_DOUBLE if self.cb_load else -1
+            cb_store_type = CUFFT_CB_ST_COMPLEX_DOUBLE if self.cb_store else -1
+        elif fft_type == CUFFT_D2Z:
+            cb_load_type = CUFFT_CB_LD_REAL_DOUBLE if self.cb_load else -1
+            cb_store_type = CUFFT_CB_ST_COMPLEX_DOUBLE if self.cb_store else -1
+        elif fft_type == CUFFT_Z2D:
+            cb_load_type = CUFFT_CB_LD_COMPLEX_DOUBLE if self.cb_load else -1
+            cb_store_type = CUFFT_CB_ST_REAL_DOUBLE if self.cb_store else -1
+        else:
+            raise ValueError
+
+        from cupy.cuda import cufft
+        cdef intptr_t plan = cufft.create()
+        cufft.setAutoAllocation(plan, 0)
+
+        if self.cb_load:
+            if cb_load_data is not None:
+                cb_load_ptr = cb_load_data.ptr
+            cufft.setJITCallback(
+                <intptr_t>plan, self.cb_load_name, self.cb_load_lto,
+                cb_load_type, cb_load_ptr)
+        if self.cb_store:
+            if cb_store_data is not None:
+                cb_store_ptr = cb_store_data.ptr
+            cufft.setJITCallback(
+                <intptr_t>plan, self.cb_store_name, self.cb_store_lto,
+                cb_store_type, cb_store_ptr)
+
+        return <intptr_t>plan
+
+
+# LHS is our legacy convention
+# RHS is a valid C identifier pattern
+cdef object cupy_callback_pattern = re.compile(
+    r'd_(?:load|store)CallbackPtr\s*=\s*([A-Za-z_]\w*)\s*;')
```

Comment:
How about separating the patterns to handle cases where users might pass arguments in the wrong order?

```suggestion
cdef object cupy_load_callback_pattern = re.compile(
    r'd_loadCallbackPtr\s*=\s*([A-Za-z_]\w*)\s*;')

cdef object cupy_store_callback_pattern = re.compile(
    r'd_storeCallbackPtr\s*=\s*([A-Za-z_]\w*)\s*;')
```